test/bundler-integration/module/HybridSessionKeyManager.Module.specs.ts

```diff
@@ -0,0 +1,555 @@
+    ).deploy();
+
+    const validUntil = 0;
+    const validAfter = 0;
+    const sessionKeyData = hexZeroPad(sessionKey.address, 20);
+    const sessionData: ISessionKeyManagerModuleHybrid.SessionDataStruct = {
+      validUntil,
+      validAfter,
+      sessionKeyData,
+      sessionValidationModule: mockSessionValidationModule1.address,
+    };
+    const sessionData2: ISessionKeyManagerModuleHybrid.SessionDataStruct = {
+      ...sessionData,
+      sessionValidationModule: mockSessionValidationModule2.address,
+    };
+
+    const hybridSKMSingleCallUtils = new HybridSKMSingleCallUtils(
+      entryPoint,
+      hybridSessionKeyManager,
+      ecdsaModule
+    );
+    const hybridSKMBatchCallUtils = new HybridSKMBatchCallUtils(
+      entryPoint,
+      hybridSessionKeyManager,
+      ecdsaModule
+    );
+
+    return {
+      entryPoint: entryPoint,
+      smartAccountImplementation: await getSmartAccountImplementation(),
+      smartAccountFactory: await getSmartAccountFactory(),
+      mockToken: mockToken,
+      ecdsaModule: ecdsaModule,
+      userSA: userSA,
+      hybridSessionKeyManager,
+      mockSessionValidationModule: mockSessionValidationModule1,
+      mockSessionValidationModule2,
+      sessionKeyData: sessionKeyData,
+      sessionData,
+      sessionData2,
+      hybridSKMSingleCallUtils,
+      hybridSKMBatchCallUtils,
+      smartAccountOwner,
+    };
+  });
+
+  describe("Single Call", async () => {
+    it("Should process signed user operation from Session when enabling and use is batched", async () => {
+      const {
+        entryPoint,
+        userSA,
+        hybridSKMSingleCallUtils: utils,
+        mockToken,
+        sessionData,
+        smartAccountOwner,
+      } = await setupTests();
+      const tokenAmountToTransfer = ethers.utils.parseEther("0.834");
+
+      const chainId = (await ethers.provider.getNetwork()).chainId;
+
+      const { sessionEnableData, sessionEnableSignature } =
+        await utils.makeSessionEnableData(
+          [chainId],
+          [sessionData],
+          userSA.address,
+          smartAccountOwner
+        );
+
+      const transferUserOp =
+        await utils.makeEcdsaSessionKeySignedUserOpForEnableAndUseSession(
+          userSA.address,
+          {
+            to: mockToken.address,
+            value: ethers.utils.parseEther("0"),
+            calldata: encodeTransfer(
+              charlie.address,
+              tokenAmountToTransfer.toString()
+            ),
+          },
+          sessionKey,
+          sessionData,
+          sessionEnableData,
+          sessionEnableSignature,
+          0,
+          {
+            preVerificationGas: 51000,
+          }
+        );
+
+      const charlieTokenBalanceBefore = await mockToken.balanceOf(
+        charlie.address
+      );
+
+      await environment.sendUserOperation(transferUserOp, entryPoint.address);
+
+      expect(await mockToken.balanceOf(charlie.address)).to.equal(
```

Comment:
Should expect the event emitted for SessionCreated 
After the transaction, query the contract to see if the session is actually enabled and cross check the count of sessions that are enabled.

Reply:
The bundler integration tests only check for compatibility with the bundler, and similar positive scenario tests are implemented in foundry that verify all the conditions you've mentioned.
Please refer to 
1. https://github.com/bcnmy/scw-contracts/blob/feat/sma-392-session-keys-v2/test/foundry/module/SessionKeyManager/SessionKeyManagerHybrid.BatchCall.t.sol#L87
2. https://github.com/bcnmy/scw-contracts/blob/feat/sma-392-session-keys-v2/test/foundry/module/SessionKeyManager/SessionKeyManagerHybrid.SingleCall.t.sol#L114

and other positive scenario tests in foundry, where event emissions and state checks have been thoroughly tested. You'll also find tests for enabling and disabling sessions there.

In general, I've implemented all positive and negative tests in foundry, and have only kept basic positive flow tests in bunder/hardhat to ensure bundler compatibility.

Reply:
The full list of tests implemented in foundry: 
```
Running 18 tests for test/foundry/module/SessionKeyManager/SessionKeyManagerHybrid.SingleCall.t.sol:SessionKeyManagerHybridSingleCallTest
[PASS] testDisableSession() (gas: 208224)
[PASS] testDisableSessions() (gas: 209324)
[PASS] testEnableAndUseSessionInSameTransaction() (gas: 206419)
[PASS] testEnableAndUseSessionMultiSessionEnable() (gas: 267433)
[PASS] testEnableAndUseSessionPostSessionEnable() (gas: 243050)
[PASS] testEnableSession() (gas: 165474)
[PASS] testEnableSessions() (gas: 236559)
[PASS] testExplicitEnableAndUseSessionDifferentOp() (gas: 216171)
[PASS] testShouldNotSupportERC1271SignatureValidation(uint256) (runs: 256, μ: 12261, ~: 12261)
[PASS] testShouldNotSupportERC1271SignatureValidationUnsafe(uint256) (runs: 256, μ: 12195, ~: 12195)
[PASS] testShouldNotValidateTransactionFromNonEnabledSession() (gas: 170546)
[PASS] testShouldNotValidateTransactionFromNonEnabledSessionWithPostCacheFlow() (gas: 119975)
[PASS] testShouldNotValidateTransactionSignedFromInvalidSessionSigner() (gas: 210961)
[PASS] testShouldNotValidateTransactionSignedFromInvalidSessionSignerPostSessionEnable() (gas: 270079)
[PASS] testShouldNotValidateTransactionWithInvalidChainId() (gas: 153673)
[PASS] testShouldNotValidateTransactionWithInvalidSessionIndex() (gas: 153168)
[PASS] testShouldParseEnableSessionSignatureCorrectly(uint8,uint8,uint48,uint48,address,bytes,bytes,bytes,bytes) (runs: 256, μ: 61973, ~: 51556)
[PASS] testShouldParsePreEnabledSignatureCorrectly(uint8,bytes32,bytes) (runs: 256, μ: 21042, ~: 18046)
Test result: ok. 18 passed; 0 failed; 0 skipped; finished in 3.74s

Running 21 tests for test/foundry/module/SessionKeyManager/SessionKeyManagerHybrid.BatchCall.t.sol:SessionKeyManagerHybridBatchCallTest
[PASS] testEnableAndUseSessionSingleBatchItem() (gas: 227729)
[PASS] testEnableAndUseSessionTwoBatchItems() (gas: 332589)
[PASS] testShouldNotAllowInvalidChainIdInSessionEnableData() (gas: 166563)
[PASS] testShouldNotAllowInvalidSessionEnableDataIndex() (gas: 166247)
[PASS] testShouldNotAllowInvalidSessionEnableKeyIndex() (gas: 166860)
[PASS] testShouldNotAllowSessionEnableWithEnableDataListLengthMismatch() (gas: 147112)
[PASS] testShouldNotAllowSessionEnableWithInvalidSessionEnableSignature() (gas: 161367)
[PASS] testShouldNotAllowSessionExecutionIfSVMReturnsDifferentSigner() (gas: 216672)
[PASS] testShouldNotAllowSessionInfosLengthAndBatchItemsLengthMismatch() (gas: 168124)
[PASS] testShouldNotAllowSessionWithDifferentDigestToBeExecuted() (gas: 169453)
[PASS] testShouldNotAllowUsageOfExpiredSession() (gas: 298389)
[PASS] testShouldNotAllowUsageOfNonEnabledSessionsInPreEnabledFlow() (gas: 132143)
[PASS] testShouldNotAllowUsageOfSessionsNotValidYet() (gas: 298728)
[PASS] testShouldNotSupportERC1271SignatureValidation(uint256) (runs: 256, μ: 12287, ~: 12287)
[PASS] testShouldNotSupportERC1271SignatureValidationUnsafe(uint256) (runs: 256, μ: 12111, ~: 12111)
[PASS] testShouldParseBatchCallDataCorrectly(address[],uint256[],bytes[]) (runs: 256, μ: 4337881, ~: 4337168)
[PASS] testShouldParseSessionDataPreEnableSignatureBatchCall(uint8,bytes32,bytes) (runs: 256, μ: 20731, ~: 17927)
[PASS] testShouldParseSessionEnableSignatureCorrectly(uint8,uint8,uint8,uint48,uint48,address,bytes,bytes) (runs: 256, μ: 35498, ~: 29535)
[PASS] testShouldParseValidateUserBatchSignature(bytes[],bytes[],bytes[],bytes) (runs: 256, μ: 13238584, ~: 13069998)
[PASS] testUseSessionTwoBatchItemsOneFreshAndOtherPostEnable() (gas: 389251)
[PASS] testUseSessionTwoBatchItemsPostEnable() (gas: 387304)
Test result: ok. 21 passed; 0 failed; 0 skipped; finished in 10.71s
```